clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -903,6 +901,65 @@ private void updateBytesBuffered(boolean madeProgress) {
 
     @Override
     public long transferFrom(FileChannel fileChannel, long position, long count) throws IOException {
-        return fileChannel.transferTo(position, count, this);
+        if (state == State.CLOSING)
+            throw closingException();
+        if (state != State.READY)
+            return 0;
+
+        if (!flush(netWriteBuffer))
+            return 0;
+
+        long channelSize = fileChannel.size();
+        if (position > channelSize)
+            return 0;
+        int bytesToWrite = (int) Math.min(Math.min(count, channelSize - position), Integer.MAX_VALUE);
+
+        if (fileChannelBuffer == null) {
+            // Pick a size that allows for reasonably efficient disk reads, keeps the memory overhead per connection
+            // manageable and can typically be drained in a single `write` call. The `netWriteBuffer` is typically 16k
+            // and the socket send buffer is 100k by default, so 32k is a good number given the mentioned trade-offs.
+            int transferSize = 32768;
+            // Allocate a direct buffer to avoid one heap to heap buffer copy. SSLEngine copies the source
+            // buffer (fileChannelBuffer) to the destination buffer (netWriteBuffer) and then encrypts in-place.
+            // FileChannel.read() to a heap buffer requires a copy from a direct buffer to a heap buffer, which is not
+            // useful here.
+            fileChannelBuffer = ByteBuffer.allocateDirect(transferSize);
+            // The loop below drains any remaining bytes from the buffer before reading from disk, so we ensure there
+            // are no remaining bytes in the empty buffer
+            fileChannelBuffer.position(fileChannelBuffer.limit());
+        }
+
+        int bytesWritten = 0;
+        long pos = position;
+        try {
+            while (bytesWritten < bytesToWrite) {
+                int bytesRead = fileChannelBuffer.remaining();
+                if (!fileChannelBuffer.hasRemaining()) {
+                    fileChannelBuffer.clear();
+                    int bytesRemaining = bytesToWrite - bytesWritten;
+                    if (bytesRemaining < fileChannelBuffer.limit())
+                        fileChannelBuffer.limit(bytesRemaining);
+                    bytesRead = fileChannel.read(fileChannelBuffer, pos);
+                    if (bytesRead <= 0)
+                        break;
+                    fileChannelBuffer.flip();
+                }
+                int networkBytesWritten = write(fileChannelBuffer);
+                bytesWritten += networkBytesWritten;
+                if (networkBytesWritten < bytesRead) {
+                    // Handle partial write by moving the remaining bytes to the start of the buffer so that we write
+                    // them first in the next transferFrom call
+                    fileChannelBuffer.compact();
+                    fileChannelBuffer.flip();
```

Comment:
The caller will request the remaining 70, but if the fileChannalBuffer still has some data, we will try to drain that first. If we can't drain it, we will return 0 and the callers will try again. If we can drain it, we still have space in the network buffer and `totalBytesWritten < totalBytesToWrite`, we will read from disk again. Does that make sense?

Reply:
@junrao explained to me what the problem is, well spotted! I pushed a commit that I think fixes it.

Reply:
I was doing additional testing and this change broke things. I looked into it and I now understand why it worked without it. We do `pos += networkBytesWritten` after we drain `fileChannelBuffer` and we only do `fileChannel.read` after draining `fileChannelBuffer`. Given that, we don't need to adjust `position` separately. Does that make sense?

Reply:
Thanks for the explanation. That makes sense. Since we always drain the remaining bytes in fileChannelBuffer first before filling in new data, do we need to do fileChannelBuffer.compact() here?